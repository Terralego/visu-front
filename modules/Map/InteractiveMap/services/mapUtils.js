export function toggleLayerVisibility (map, layerId, visibility) {
  map.setLayoutProperty(layerId, 'visibility', visibility);
}

export function getOpacityProperty (type) {
  switch (type) {
    case 'background':
      return 'background-opacity';
    case 'fill':
      return 'fill-opacity';
    case 'line':
      return 'line-opacity';
    case 'symbol':
      return 'icon-opacity';
    case 'raster':
      return 'raster-opacity';
    case 'circle':
      return 'circle-opacity';
    case 'fill-extrusion':
      return 'fill-extrusion-opacity';
    default:
      return null;
  }
}

export function setLayerOpacity (map, layerId, opacity) {
  const layer = map.getLayer(layerId);
  const property = getOpacityProperty(layer.type);
  if (property) {
    map.setPaintProperty(layerId, property, opacity);
  }
}

export function getInteractionOnEvent ({ eventType, map, point, interactions }) {
  const features = map.queryRenderedFeatures(point);
  let interaction = false;

  features.some(feature => {
    const { layer: { id: layerId } } = feature;

    const foundInteraction = interactions.find(({ id, trigger = 'click' }) => {
      const found = id === layerId;

      return found &&
        eventType === (trigger === 'mouseover' ? 'mousemove' : trigger);
    });

    if (!foundInteraction) return false;

    interaction = {
      interaction: foundInteraction,
      feature,
      layerId,
    };

    return true;
  });

  return interaction;
}

export function setInteractions ({ map, interactions, callback }) {
  const eventsTypes = new Set(interactions.reduce((triggers, { trigger = 'click' }) => [...triggers, trigger], []));

  if (eventsTypes.has('mouseover')) {
    eventsTypes.add('mousemove');
    eventsTypes.delete('mouseover');
  }

  eventsTypes.forEach(eventType => {
    map.on(eventType, e => {
      const { target, point } = e;
      const interactionSpec = getInteractionOnEvent({
        eventType,
        map: target,
        point,
        interactions,
      });

      if (!interactionSpec) return;

      const { interaction, feature, layerId } = interactionSpec;

      callback({ event: e, map, layerId, feature, interaction, eventType });
    });
  });

  /**
   * Mouseleave events for mouseover triggers
   */
  interactions.forEach(interaction => {
    const { id, trigger } = interaction;
    if (trigger !== 'mouseover') return;

    const eventType = 'mouseleave';
    map.on(eventType, id, event => {
      callback({ event, map, layerId: id, interaction, eventType });
    });
  });

  /**
   *  Display a pointer cursor over click zones
   */
  map.on('mousemove', e => {
    const { target, point } = e;
    const interactionSpec = getInteractionOnEvent({
      eventType: 'click',
      map: target,
      point,
      interactions,
    });

    const canvas = target.getCanvas();
    if (interactionSpec) {
      canvas.style.cursor = 'pointer';
    } else {
      canvas.style.cursor = '';
    }
  });
}

export const checkContraints = ({
  map,
  constraints = [],
  feature: { properties: { cluster } = {} } = {},
}) => {
  if (!constraints.length) return true;
  const currentZoom = map.getZoom();
  return constraints.reduce((prev, {
    minZoom = 0,
    maxZoom = Infinity,
    withLayers = [],
    isCluster,
  }) => {
    const checkZoom = minZoom !== undefined && maxZoom !== undefined
      ? currentZoom <= maxZoom && currentZoom >= minZoom
      : true;

    const checkLayers = withLayers.reduce((prevCheck, layer) => {
      const match = layer.match(/^(!)?(.+)$/);

      const visible = match[1] !== '!';
      const layerId = match[2];

      return prevCheck
        && visible
        ? (map.getLayer(layerId) !== undefined
          && map.getLayoutProperty(layerId, 'visibility') === 'visible')
        : (!map.getLayer(layerId)
          || map.getLayoutProperty(layerId, 'visibility') === 'none');
    }, true);
    const checkCluster = isCluster === undefined
      || cluster === isCluster;

    return prev || (checkZoom && checkLayers && checkCluster);
  }, false);
};

export default {
  toggleLayerVisibility,
  getOpacityProperty,
  setLayerOpacity,
  getInteractionOnEvent,
  setInteractions,
  checkContraints,
};
