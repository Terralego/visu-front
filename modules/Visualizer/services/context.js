import React from 'react';

import connect from '../../../utils/connect';

const context = React.createContext();
export const { Provider, Consumer } = context;

export const connectVisualizerProvider = connect(context);

export default context;
