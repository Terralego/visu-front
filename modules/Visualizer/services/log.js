import debug from 'debug';

export const log = debug('terralego:Visualizer');

export default log;
