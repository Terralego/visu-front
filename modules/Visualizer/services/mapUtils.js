export function toggleLayerVisibility (map, layerId, visibility) {
  map.setLayoutProperty(layerId, 'visibility', visibility);
}

export function getOpacityProperty (type) {
  switch (type) {
    case 'background':
      return 'background-opacity';
    case 'fill':
      return 'fill-opacity';
    case 'line':
      return 'line-opacity';
    case 'symbol':
      return 'icon-opacity';
    case 'raster':
      return 'raster-opacity';
    case 'circle':
      return 'circle-opacity';
    case 'fill-extrusion':
      return 'fill-extrusion-opacity';
    default:
      return null;
  }
}

export function setLayerOpacity (map, layerId, opacity) {
  const layer = map.getLayer(layerId);
  const property = getOpacityProperty(layer.type);
  if (property) {
    map.setPaintProperty(layerId, property, opacity);
  }
}

export function getInteractionOnEvent ({ eventType, map, point, interactions }) {
  const features = map.queryRenderedFeatures(point);
  let interaction = false;

  features.some(feature => {
    const { layer: { id: layerId } } = feature;

    const foundInteraction = interactions.find(({ id, trigger = 'click' }) => {
      const found = id === layerId;

      return found &&
        eventType === (trigger === 'mouseover' ? 'mousemove' : trigger);
    });

    if (!foundInteraction) return false;

    interaction = {
      interaction: foundInteraction,
      feature,
      layerId,
    };

    return true;
  });

  return interaction;
}

export function setInteractions ({ map, interactions, callback }) {
  const eventsTypes = new Set(interactions.reduce((triggers, { trigger = 'click' }) => [...triggers, trigger], []));

  if (eventsTypes.has('mouseover')) {
    eventsTypes.add('mousemove');
    eventsTypes.delete('mouseover');
  }

  eventsTypes.forEach(eventType => {
    map.on(eventType, e => {
      const { target, point } = e;
      const interactionSpec = getInteractionOnEvent({
        eventType,
        map: target,
        point,
        interactions,
      });

      if (!interactionSpec) return;

      const { interaction, feature, layerId } = interactionSpec;

      callback({ event: e, map, layerId, feature, interaction, eventType });
    });
  });

  /**
   * Mouseleave events for mouseover triggers
   */
  interactions.forEach(interaction => {
    const { id, trigger } = interaction;
    if (trigger !== 'mouseover') return;

    const eventType = 'mouseleave';
    map.on(eventType, id, event => {
      callback({ event, map, layerId: id, interaction, eventType });
    });
  });

  /**
   *  Display a pointer cursor over click zones
   */
  map.on('mousemove', e => {
    const { target, point } = e;
    const interactionSpec = getInteractionOnEvent({
      eventType: 'click',
      map: target,
      point,
      interactions,
    });

    const canvas = target.getCanvas();
    if (interactionSpec) {
      canvas.style.cursor = 'pointer';
    } else {
      canvas.style.cursor = '';
    }
  });
}

export default {
  toggleLayerVisibility,
  getOpacityProperty,
  setLayerOpacity,
  getInteractionOnEvent,
  setInteractions,
};
