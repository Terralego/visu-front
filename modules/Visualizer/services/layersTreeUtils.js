import { PREFIX_SOURCE } from '../../Map/services/cluster';

export const INITIAL_FILTERS = new Map();

export const isCluster = (source, layerId) => !!source.match(new RegExp(`^${layerId}-${PREFIX_SOURCE}-[0-9]+`));

export function initLayersStateAction (layersTree) {
  const layersTreeState = new Map();
  function reduceLayers (group, map) {
    return group.reduce((layersStateMap, layer) => {
      const { initialState = {}, sublayers } = layer;
      const { active } = initialState;
      if (sublayers) {
        initialState.sublayers = initialState.sublayers || sublayers.map((_, k) =>
          (k === 0 && !!active));
      }
      if (layer.group) {
        return reduceLayers(layer.layers, layersStateMap);
      }
      initialState.opacity = initialState.opacity === undefined
        ? 1
        : initialState.opacity;
      layersStateMap.set(layer, {
        active: false,
        opacity: 1,
        ...initialState,
      });
      return layersStateMap;
    }, map);
  }
  return reduceLayers(layersTree, layersTreeState);
}

export function setLayerStateAction (layer, layerState, prevLayersTreeState) {
  const layersTreeState = new Map(prevLayersTreeState);
  const prevLayerState = layersTreeState.get(layer);
  const newLayerState = { ...layerState };

  if (!prevLayerState) return prevLayersTreeState;

  if (prevLayerState.sublayers && !prevLayerState.sublayers.find(sl => sl)) {
    newLayerState.sublayers = [...prevLayerState.sublayers];
    newLayerState.sublayers[0] = true;
  }
  if (newLayerState.table) {
    // Easiest to to read as transform Map in Array and run a .map() on it
    // eslint-disable-next-line no-param-reassign
    layersTreeState.forEach(layState => { layState.table = false; });
  }
  layersTreeState.set(layer, {
    ...prevLayerState,
    ...newLayerState,
  });
  return layersTreeState;
}

export function layersTreeStatesHaveChanged (layersTreeState, prevLayersTreeState, fields = ['active']) {
  return !fields.reduce((all, field) =>
    all && Array.from(layersTreeState).reduce((suball, [layer, state = {}]) =>
      suball && state[field] === (prevLayersTreeState.get(layer) || {})[field],
    all),
  true);
}

export function selectSublayerAction (layer, sublayer, prevLayersTreeState) {
  const layersTreeState = new Map(prevLayersTreeState);
  const layerState = layersTreeState.get(layer);
  layerState.sublayers = layerState.sublayers.map((_, k) => k === sublayer);
  layersTreeState.set(layer, { ...layerState });
  return layersTreeState;
}

export const filterLayersStatesFromLayersState = (
  layersTreeState = [],
  check = ({ active }) => !!active,
) => Array
  .from(layersTreeState)
  .reduce((allLayers, [layer, state]) =>
    ((check(state)) ? [...allLayers, [layer, state]] : allLayers), []);

export const filterLayersFromLayersState = (layersTreeState, check) =>
  filterLayersStatesFromLayersState(layersTreeState, check)
    .filter(([{ filters: { layer } = {} }]) => layer)
    .map(([{ filters: { layer } }]) => layer);

export const hasTable = layersTreeState =>
  filterLayersFromLayersState(layersTreeState, ({ table }) => table).length > 0;

export const hasWidget = layersTreeState =>
  filterLayersStatesFromLayersState(layersTreeState, ({ widgets = [] }) => widgets.length > 0)
    .length > 0;

export const filterFeatures = (
  map,
  features/* = [ { layer: String, features: [id1, id2, â€¦] } ] */,
  layersTreeState,
) => {
  Array.from(layersTreeState).forEach(([{
    layers, filters: { layer } = {},
  }, {
    active,
  }]) => {
    if (!active || !layers) return;
    const layerFeatures = features
      .find(({ layer: fLayer }) => fLayer === layer);
    const { features: ids = [] } = layerFeatures || {};

    const filter = ids.length ? ['match', ['get', '_id'], ids, true, false] : null;
    layers.forEach(layerId => {
      const paintLayer = map.getLayer(layerId);
      if (!paintLayer) return;

      if (isCluster(paintLayer.source, layerId)) {
        // Force to upgrade cluster data
        map.fire('refreshCluster');
        return;
      }

      if (!INITIAL_FILTERS.has(layerId)) {
        INITIAL_FILTERS.set(layerId, map.getFilter(layerId));
      }
      map.setFilter(layerId, layerFeatures ? filter : INITIAL_FILTERS.get(layerId));
    });
  });
};

export const resetFilters = (map, layersTreeState) => {
  Array.from(layersTreeState).forEach(([{
    layers = [], filters,
  }]) => {
    if (!filters) return;

    layers.forEach(layerId => {
      const paintLayer = map.getLayer(layerId);

      if (!map.getLayer(layerId)) return;

      if (isCluster(paintLayer.source, layerId)) {
        // Force to upgrade cluster data
        map.fire('refreshCluster');
        return;
      }
      if (!INITIAL_FILTERS.has(layerId)) return;
      map.setFilter(layerId, INITIAL_FILTERS.get(layerId));
    });
  });
};

export default {
  initLayersStateAction,
  selectSublayerAction,
  setLayerStateAction,
  filterLayersFromLayersState,
  hasTable,
  hasWidget,
  filterFeatures,
  resetFilters,
};
