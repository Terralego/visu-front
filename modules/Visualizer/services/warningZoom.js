import { getRelatedLayers } from '../../Map/services/mapUtils';

export function displayWarningAccordingToZoom (map, layer) {
  if (!map) return {};
  const currentZoom = map.getZoom();
  const layers = layer.layers.reduce((prev, layerId) =>
    [...prev, ...getRelatedLayers(map, layerId)], []);
  const layerZoom = layers
    .reduce(({ minzoom: prevMinzoom, maxzoom: prevMaxzoom }, { id }) => {
      const { minzoom = 0, maxzoom = 24 } = map.getLayer(id) || {};
      return {
        minzoom: Math.min(minzoom, prevMinzoom),
        maxzoom: Math.max(maxzoom, prevMaxzoom),
      };
    }, { minzoom: 24, maxzoom: 0 });
  const isNotDisplayed = currentZoom < layerZoom.minzoom || currentZoom > layerZoom.maxzoom;

  return { display: isNotDisplayed, minZoomLayer: layerZoom.minzoom };
}

export default { displayWarningAccordingToZoom };
