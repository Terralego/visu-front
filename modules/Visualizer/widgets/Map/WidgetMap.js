import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import mapBoxGl from 'mapbox-gl';
import debounce from 'lodash.debounce';

import { context } from './connect';
import { toggleLayerVisibility, addListenerOnLayer, setLayerOpacity } from '../../services/mapUtils';
import LayersTreeProps from '../../propTypes/LayersTreePropTypes';
import DefaultMapComponent from './components/Map';
import LayersTree from './components/LayersTree';
import BackgroundStyles from './components/BackgroundStyles';
import MarkdownRenderer from '../../components/MarkdownRenderer';
import Legend from './components/Legend';

import './styles.scss';

export const INTERACTION_DISPLAY_DETAILS = 'displayDetails';
export const INTERACTION_DISPLAY_TOOLTIP = 'displayTooltip';
export const INTERACTION_FN = 'function';

const { Provider } = context;

export class WidgetMap extends React.Component {
  static propTypes = {
    layersTree: LayersTreeProps,
    backgroundStyle: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.objectOf(PropTypes.string)),
      PropTypes.string,
    ]),
    interactions: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.string.isRequired,
      trigger: PropTypes.oneOf(['click', 'mouseover']),
      interaction: PropTypes.oneOf([
        INTERACTION_DISPLAY_DETAILS,
        INTERACTION_DISPLAY_TOOLTIP,
        INTERACTION_FN,
      ]),
      // for INTERACTION_DISPLAY_DETAILS
      // and INTERACTION_DISPLAY_TOOLTIP
      template: PropTypes.string,
      content: PropTypes.string,
      // for INTERACTION_FN
      fn: PropTypes.func,
    })),
    LayersTreeComponent: PropTypes.func,
    MapComponent: PropTypes.func,
    setDetails: PropTypes.func,
  };

  static defaultProps = {
    backgroundStyle: 'mapbox://styles/mapbox/light-v9',
    layersTree: [],
    LayersTreeComponent: LayersTree,
    MapComponent: DefaultMapComponent,
    interactions: [],
    setDetails () {},
  };

  mapRef = React.createRef();

  state = {
    selectedBackgroundStyle: Array.isArray(this.props.backgroundStyle)
      ? this.props.backgroundStyle[0].url
      : this.props.backgroundStyle,
    layersTreeState: new Map(),
    isLayersTreeVisible: true,
  };

  map = new Promise(resolve => {
    const waitForMap = () => {
      if (this.mapRef && this.mapRef.current && this.mapRef.current.map) {
        resolve(this.mapRef.current.map);
        return;
      }
      setTimeout(waitForMap, 10);
    };
    waitForMap();
  });

  popups = new Map();

  hideTooltip = debounce(({ layerId }) => {
    if (!this.popups.has(layerId)) {
      return;
    }
    const popup = this.popups.get(layerId);
    popup.remove();
    this.popups.delete(layerId);
  }, 100);

  componentDidMount () {
    this.initLayersState();
    this.setInteractions();
  }

  componentDidUpdate ({
    interactions: prevInteractions,
  }, {
    layersTreeState: prevLayersTreeState,
  }) {
    const { interactions } = this.props;
    const { layersTreeState } = this.state;

    if (interactions !== prevInteractions) {
      this.setInteractions();
    }

    if (prevLayersTreeState !== layersTreeState) {
      this.updateLayersTree();
    }
  }

  onBackgroundChange = selectedBackgroundStyle => this.setState({ selectedBackgroundStyle });

  get legends () {
    const { layersTreeState } = this.state;
    const legends = Array.from(layersTreeState.entries())
      .map(([layer, state]) => {
        if (!state.active) return undefined;
        if (layer.sublayers) {
          const selected = state.sublayers.findIndex(active => active);
          const selectedSublayer = layer.sublayers[selected];
          return selectedSublayer && selectedSublayer.legend && ({
            title: selectedSublayer.label,
            ...selectedSublayer.legend,
          });
        }
        return layer.legend && ({
          title: layer.label,
          ...layer.legend,
        });
      })
      .filter(defined => defined);

    return legends;
  }

  async setInteractions () {
    const map = await this.map;
    const { interactions } = this.props;

    if (this.mapInteractionsListeners) {
      this.mapInteractionsListeners.forEach(off => off());
    }

    this.mapInteractionsListeners = [];

    interactions.forEach(({ id, trigger = 'click', interaction, fn, ...config }) => {
      const calls = [];
      switch (interaction) {
        case INTERACTION_DISPLAY_DETAILS:
          calls.push({
            callback: (layerId, features) => this.displayDetails({ features, ...config }),
            trigger,
            displayCursor: trigger === 'click',
          });
          break;
        case INTERACTION_DISPLAY_TOOLTIP:
          if (trigger === 'mouseover') {
            calls.push({
              callback: (layerId, features, event) =>
                this.displayTooltip({ layerId, features, event, ...config }),
              trigger: 'mousemove',
            });
            calls.push({
              callback: (layerId, features, event) =>
                this.hideTooltip({ layerId, features, event, ...config }),
              trigger: 'mouseleave',
            });
          } else {
            calls.push({
              callback: (layerId, features, event) =>
                this.displayTooltip({ layerId, features, event, ...config }),
              trigger,
              displayCursor: trigger === 'click',
            });
          }
          break;
        case INTERACTION_FN:
          calls.push({
            callback: fn,
            trigger,
            displayCursor: trigger === 'click',
          });
          break;
        default:
          return;
      }
      const listeners = calls.reduce((list, { callback, trigger: callTrigger, displayCursor }) => [
        ...list,
        ...addListenerOnLayer(
          map,
          id,
          callback,
          { displayCursor, trigger: callTrigger },
        ),
      ], []);

      this.mapInteractionsListeners = this.mapInteractionsListeners.concat(listeners);
    });
  }

  setLayerState = ({ layer, state }) => {
    const { layersTreeState: prevLayersTreeState } = this.state;
    const layersTreeState = new Map(prevLayersTreeState);
    const layerState = layersTreeState.get(layer);

    if (!layerState) return;

    layersTreeState.set(layer, {
      ...layerState,
      ...state,
    });
    this.setState({ layersTreeState });
  }

  getLayerState = ({ layer }) => {
    const { layersTreeState } = this.state;
    const layerState = layersTreeState.get(layer);

    return layerState || {};
  }

  displayDetails = ({ features, template }) => {
    const { setDetails } = this.props;
    setDetails({ features, template });
  }

  displayTooltip = async ({
    layerId,
    features: [{ properties }] = [{}],
    event: { lngLat },
    template,
    content,
  }) => {
    const map = await this.map;
    const container = document.createElement('div');
    ReactDOM.render(
      <MarkdownRenderer
        template={template}
        content={content}
        {...properties}
      />,
      container,
    );

    if (this.popups.has(layerId)) {
      this.popups.get(layerId).setLngLat([lngLat.lng, lngLat.lat]);
      return;
    }
    const popup = new mapBoxGl.Popup();
    this.popups.set(layerId, popup);
    popup.setLngLat([lngLat.lng, lngLat.lat]);
    popup.setDOMContent(container);
    popup.addTo(map);
  }

  selectSublayer = ({ layer, sublayer }) => {
    const { layersTreeState: prevLayersTreeState } = this.state;
    const layersTreeState = new Map(prevLayersTreeState);
    const layerState = layersTreeState.get(layer);
    layerState.sublayers = layerState.sublayers.map((_, k) => k === sublayer);
    layersTreeState.set(layer, { ...layerState });
    this.setState({ layersTreeState });
  }

  toggleLayersTree = () => this.setState({ isLayersTreeVisible: !this.state.isLayersTreeVisible })

  initLayersState () {
    const { layersTree } = this.props;
    const layersTreeState = new Map();
    function reduceLayers (group, map) {
      return group.reduce((layersStateMap, layer) => {
        const { initialState = {}, sublayers } = layer;
        const { active } = initialState;
        if (sublayers) {
          initialState.sublayers = initialState.sublayers || sublayers.map((_, k) =>
            (k === 0 && !!active));
        }
        if (layer.group) {
          return reduceLayers(layer.layers, layersStateMap);
        }
        initialState.opacity = initialState.opacity || 1;
        layersStateMap.set(layer, {
          active: false,
          opacity: 1,
          ...initialState,
        });
        return layersStateMap;
      }, map);
    }
    this.setState({
      layersTreeState: reduceLayers(layersTree, layersTreeState),
    });
  }

  async updateLayersTree () {
    const map = await this.map;
    const { layersTreeState } = this.state;
    layersTreeState.forEach(({ active, opacity, sublayers }, layer) => {
      if (sublayers) {
        if (active) {
          layer.sublayers.forEach((sublayer, index) => {
            const sublayerActive = sublayers[index];
            sublayer.layers.forEach(layerId => {
              toggleLayerVisibility(map, layerId, sublayerActive ? 'visible' : 'none');
            });
          });
        } else {
          layer.sublayers.forEach(sublayer => {
            sublayer.layers.forEach(layerId => {
              toggleLayerVisibility(map, layerId, 'none');
            });
          });
        }

        layer.sublayers.forEach(sublayer => {
          sublayer.layers.forEach(layerId => {
            setLayerOpacity(map, layerId, opacity);
          });
        });
      }

      if (active !== undefined && layer.layers) {
        layer.layers.forEach(layerId => {
          toggleLayerVisibility(map, layerId, active ? 'visible' : 'none');
        });
      }
      if (opacity !== undefined && layer.layers) {
        layer.layers.forEach(layerId =>
          setLayerOpacity(map, layerId, opacity));
      }
    });
  }

  render () {
    const {
      LayersTreeComponent,
      MapComponent,
      layersTree,
      style,
      interactions,
      backgroundStyle,
      ...mapProps
    } = this.props;

    const { selectedBackgroundStyle, isLayersTreeVisible } = this.state;
    const {
      onChange,
      mapRef,
      getLayerState,
      setLayerState,
      selectSublayer,
      legends,
      onBackgroundChange,
      toggleLayersTree,
    } = this;
    const contextValue = {
      getLayerState,
      setLayerState,
      selectSublayer,
    };

    return (
      <Provider value={contextValue}>
        <div
          className={`widget-map ${isLayersTreeVisible ? 'widget-map--layerstree-is-visible' : ''}`}
          style={style}
        >
          {typeof backgroundStyle !== 'string' && (
            <BackgroundStyles
              onChange={onBackgroundChange}
              styles={backgroundStyle}
              selected={selectedBackgroundStyle}
            />
          )}
          {!!layersTree.length && (
          <LayersTreeComponent
            layersTree={layersTree}
            onChange={onChange}
            toggleLabel={isLayersTreeVisible ? 'replier ' : 'déplier'}
            onToggle={toggleLayersTree}
          />
          )}
          <MapComponent
            {...mapProps}
            ref={mapRef}
            backgroundStyle={selectedBackgroundStyle}
            onDetailsChange={this.onDetailsChange}
          />
          {!!legends.length && (
            <div className="widget-map__legends">
              {legends
                .map(legend => (
                  <Legend
                    key={legend.title}
                    {...legend}
                  />
                ))}
            </div>
          )}
        </div>
      </Provider>
    );
  }
}

export default WidgetMap;
