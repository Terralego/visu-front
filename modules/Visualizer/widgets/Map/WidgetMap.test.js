import React from 'react';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import mapboxGl from 'mapbox-gl';

import { toggleLayerVisibility } from '../../services/mapUtils';
import WidgetMap, { INTERACTION_DISPLAY_DETAILS, INTERACTION_DISPLAY_TOOLTIP, INTERACTION_FN } from './WidgetMap';


jest.mock('mapbox-gl', () => {
  function Popup () {}
  Popup.prototype.setLngLat = jest.fn();
  Popup.prototype.setDOMContent = jest.fn();
  Popup.prototype.addTo = jest.fn();
  return {
    Popup,
  };
});
jest.mock('./components/Map', () => {
  // A class component is needed to accept the ref
  // eslint-disable-next-line global-require
  const { Component } = require('react');
  class MapComponent extends Component {
    state = {};
    render () {
      return null;
    }
  }
  return MapComponent;
});
jest.mock('../../services/mapUtils', () => ({
  toggleLayerVisibility: jest.fn(),
  addListenerOnLayer: jest.fn((map, id, callback) => [{ map, id, callback }]),
}));
jest.mock('react-dom', () => {
  const mockedElement = {};
  return {
    mockedElement,
    render: jest.fn((jsx, el) => el),
  };
});
jest.mock('lodash.debounce', () => fn => () => fn({ layerId: 'foo' }));

it('should render correctly', () => {
  const tree = renderer.create((
    <WidgetMap
      layersTree={[]}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});
it('should render correctly with a layers tree', () => {
  const tree = renderer.create((
    <WidgetMap
      layersTree={[{
        label: 'foo',
      }]}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should setInteraction on mount', () => {
  const instance = new WidgetMap();
  instance.setInteractions = jest.fn();
  instance.componentDidMount();
  expect(instance.setInteractions).toHaveBeenCalled();
});

it('should setInteraction on update', () => {
  const instance = new WidgetMap({});
  instance.setInteractions = jest.fn();
  instance.componentDidUpdate({});
  expect(instance.setInteractions).not.toHaveBeenCalled();
  instance.componentDidUpdate({ interactions: {} });
  expect(instance.setInteractions).toHaveBeenCalled();
});

it('should apply changes', async done => {
  const instance = new WidgetMap();
  const map = {};
  instance.map = map;
  instance.onChange({
    layer: {
      id: 'foo',
      layers: ['bar', 'babar'],
    },
    state: {
      active: true,
    },
  });
  await true;
  expect(toggleLayerVisibility).toHaveBeenCalledWith(map, 'bar', 'visible');
  expect(toggleLayerVisibility).toHaveBeenCalledWith(map, 'babar', 'visible');
  toggleLayerVisibility.mockClear();

  instance.onChange({
    layer: {
      id: 'foo',
      layers: ['bar'],
    },
    state: {
      active: false,
    },
  });
  await true;
  expect(toggleLayerVisibility).toHaveBeenCalledWith(map, 'bar', 'none');
  expect(toggleLayerVisibility).not.toHaveBeenCalledWith(map, 'babar', 'none');
  toggleLayerVisibility.mockClear();

  instance.onChange({
    layer: {
      id: 'bar',
    },
    state: {
      something: true,
    },
  });
  await true;

  expect(toggleLayerVisibility).not.toHaveBeenCalled();
  toggleLayerVisibility.mockClear();

  done();
});

it('should set interactions', async done => {
  const customFn = jest.fn();
  const instance = new WidgetMap({
    interactions: [{
      id: 'foo',
      interaction: INTERACTION_DISPLAY_DETAILS,
    }, {
      id: 'bar',
      interaction: INTERACTION_DISPLAY_TOOLTIP,
    }, {
      id: 'foo',
      interaction: INTERACTION_DISPLAY_TOOLTIP,
      trigger: 'mouseover',
    }, {
      id: 'foo',
      interaction: INTERACTION_FN,
      fn: customFn,
    }, {
      id: 'foo',
      interaction: 'invalid interaction',
    }],
  });
  instance.map = {};
  instance.displayDetails = jest.fn();
  instance.displayTooltip = jest.fn();
  instance.hideTooltip = jest.fn();

  instance.setInteractions();
  await true;

  const { mapInteractionsListeners } = instance;
  expect(mapInteractionsListeners.length).toBe(5);
  mapInteractionsListeners[0].callback();
  expect(instance.displayDetails).toHaveBeenCalled();
  instance.displayTooltip.mockClear();
  mapInteractionsListeners[1].callback();
  expect(instance.displayTooltip).toHaveBeenCalled();
  instance.displayTooltip.mockClear();
  mapInteractionsListeners[2].callback();
  expect(instance.displayTooltip).toHaveBeenCalled();
  instance.displayTooltip.mockClear();
  mapInteractionsListeners[3].callback();
  expect(instance.hideTooltip).toHaveBeenCalled();
  instance.displayTooltip.mockClear();
  mapInteractionsListeners[4].callback();
  expect(customFn).toHaveBeenCalled();
  done();
});

it('should stop interactions', async done => {
  const instance = new WidgetMap({
    interactions: [],
  });
  const mapInteractionsListeners = [jest.fn(), jest.fn(), jest.fn()];
  instance.mapInteractionsListeners = mapInteractionsListeners;
  instance.map = {};
  instance.setInteractions();
  await true;
  expect(mapInteractionsListeners[0]).toHaveBeenCalled();
  expect(mapInteractionsListeners[1]).toHaveBeenCalled();
  expect(mapInteractionsListeners[2]).toHaveBeenCalled();
  expect(instance.mapInteractionsListeners).toEqual([]);
  done();
});

it('should wait for map loading', async done => {
  const instance = new WidgetMap();
  const { map } = instance;
  let resolved = false;
  const mapObject = {};
  map.then(awaitedMapObject => {
    resolved = awaitedMapObject;
  });
  expect(map instanceof Promise).toBe(true);
  await new Promise(resolve => setTimeout(resolve, 11));
  expect(resolved).toBe(false);
  instance.mapRef = {
    current: {
      map: mapObject,
    },
  };
  await new Promise(resolve => setTimeout(resolve, 11));
  expect(resolved).toBe(mapObject);
  done();
});

it('should display details', () => {
  const setDetails = jest.fn();
  const instance = new WidgetMap({ setDetails });
  const params = { features: {}, template: {} };
  instance.displayDetails(params);
  expect(setDetails).toHaveBeenCalledWith({ features: params.features, template: params.template });
});

it('should display tooltips', async done => {
  const instance = new WidgetMap();
  instance.map = {};
  instance.displayTooltip({
    layerId: 'foo',
    event: {
      lngLat: { lng: 1, lat: 2 },
    },
    template: 'bar',
  });
  await true;
  expect(ReactDOM.render).toHaveBeenCalled();
  expect(mapboxGl.Popup.prototype.setLngLat).toHaveBeenCalledWith([1, 2]);
  expect(mapboxGl.Popup.prototype.setDOMContent).toHaveBeenCalled();
  expect(mapboxGl.Popup.prototype.addTo).toHaveBeenCalledWith(instance.map);
  mapboxGl.Popup.prototype.setLngLat.mockClear();
  mapboxGl.Popup.prototype.setDOMContent.mockClear();
  mapboxGl.Popup.prototype.addTo.mockClear();

  instance.displayTooltip({
    layerId: 'foo',
    features: [{
      properties: {},
    }],
    event: {
      lngLat: { lng: 3, lat: 4 },
    },
    template: 'bar',
  });
  await true;
  expect(mapboxGl.Popup.prototype.setLngLat).toHaveBeenCalledWith([3, 4]);
  expect(mapboxGl.Popup.prototype.setDOMContent).not.toHaveBeenCalled();
  expect(mapboxGl.Popup.prototype.addTo).not.toHaveBeenCalled();

  done();
});

it('should hide tooltip', () => {
  const instance = new WidgetMap();
  const popup = {
    remove: jest.fn(),
  };
  instance.popups.set('foo', popup);
  instance.hideTooltip({ layerId: 'foo' });
  expect(popup.remove).toHaveBeenCalled();
  popup.remove.mockClear();
  instance.hideTooltip({ layerId: 'foo' });
  expect(popup.remove).not.toHaveBeenCalled();
});
