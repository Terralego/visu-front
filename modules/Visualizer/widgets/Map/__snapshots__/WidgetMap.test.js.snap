// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snaphsots should render background style selector 1`] = `
<div
  className="widget-map widget-map--layerstree-is-visible"
>
  <p>
    BackgroundStyles
  </p>
</div>
`;

exports[`snaphsots should render correctly 1`] = `
<div
  className="widget-map widget-map--layerstree-is-visible"
/>
`;

exports[`snaphsots should render correctly with a layers tree 1`] = `
<div
  className="widget-map widget-map--layerstree-is-visible"
>
  <p>
    LayersTree
  </p>
</div>
`;

exports[`snaphsots should render legends 1`] = `
<div
  className="widget-map widget-map--layerstree-is-visible"
>
  <p>
    LayersTree
  </p>
  <div
    className="widget-map__legends"
  >
    <div
      className="tf-legend"
    >
      <h4
        className="tf-legend__title"
      >
        foo
      </h4>
    </div>
  </div>
</div>
`;
